src/main/java/org/opensearch/ad/task/ADTaskCacheManager.java

```diff
@@ -833,6 +833,15 @@ public boolean isHCTaskCoordinatingNode(String detectorId) {
     public void setTopEntityCount(String detectorId, Integer count) {
         ADHCBatchTaskCache hcTaskCache = getExistingHCTaskCache(detectorId);
         hcTaskCache.setTopEntityCount(count);
+        ADTaskSlotLimit adTaskSlotLimit = detectorTaskSlotLimit.get(detectorId);
+
+        if (count != null && adTaskSlotLimit != null) {
+            Integer detectorTaskSlots = adTaskSlotLimit.getDetectorTaskSlots();
+            if (detectorTaskSlots != null && detectorTaskSlots > count) {
+                logger.debug("Scale down task slots from {} to the same as top entity count {}", detectorTaskSlots, count);
+                adTaskSlotLimit.setDetectorTaskSlots(count);
+            }
+        }
```

Comment:
Is this to fix some bug?

Reply:
Not a bug, this is to minimize the task slot usage. For example, we assign 10 task slots to HC detector, but actually the detector only has 3 top entities, then we should scale down task slots assigned to this HC from 10 to 3, so we can release 7 task slots for other detector's historical analysis.